rpc/client/client_test.go

```diff
@@ -214,6 +214,93 @@ func TestGetBlock(t *testing.T) {
 	require.NoError(err)
 }
 
+func TestUnconfirmedTxs(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	mockApp, rpc := getRPC(t)
+	mockApp.On("BeginBlock", mock.Anything).Return(abci.ResponseBeginBlock{})
+	mockApp.On("CheckTx", mock.Anything).Return(abci.ResponseCheckTx{})
+
+	err := rpc.node.Start()
+	require.NoError(err)
+
+	numRes, err := rpc.NumUnconfirmedTxs(context.Background())
+	assert.NoError(err)
+	assert.NotNil(numRes)
+	assert.EqualValues(0, numRes.Count)
+	assert.EqualValues(0, numRes.Total)
+	assert.EqualValues(0, numRes.TotalBytes)
+
+	limit := -1
+	txRes, err := rpc.UnconfirmedTxs(context.Background(), &limit)
+	assert.NoError(err)
+	assert.NotNil(txRes)
+	assert.EqualValues(0, txRes.Count)
+	assert.EqualValues(0, txRes.Total)
+	assert.EqualValues(0, txRes.TotalBytes)
+	assert.Empty(txRes.Txs)
+
+	tx1 := tmtypes.Tx("tx1")
+	tx2 := tmtypes.Tx("tx 2")
+
+	res, err := rpc.BroadcastTxAsync(context.Background(), tx1)
+	assert.NoError(err)
+	assert.NotNil(res)
+
+	numRes, err = rpc.NumUnconfirmedTxs(context.Background())
+	assert.NoError(err)
+	assert.NotNil(numRes)
+	assert.EqualValues(1, numRes.Count)
+	assert.EqualValues(1, numRes.Total)
+	assert.EqualValues(len(tx1), numRes.TotalBytes)
+
+	txRes, err = rpc.UnconfirmedTxs(context.Background(), &limit)
+	assert.NoError(err)
+	assert.NotNil(txRes)
+	assert.EqualValues(1, txRes.Count)
+	assert.EqualValues(1, txRes.Total)
+	assert.EqualValues(len(tx1), txRes.TotalBytes)
+	assert.Len(txRes.Txs, 1)
+	assert.Contains(txRes.Txs, tx1)
+
+	res, err = rpc.BroadcastTxAsync(context.Background(), tx2)
+	assert.NoError(err)
+	assert.NotNil(res)
+
+	numRes, err = rpc.NumUnconfirmedTxs(context.Background())
+	assert.NoError(err)
+	assert.NotNil(numRes)
+	assert.EqualValues(2, numRes.Count)
+	assert.EqualValues(2, numRes.Total)
+	assert.EqualValues(len(tx1)+len(tx2), numRes.TotalBytes)
+
+	txRes, err = rpc.UnconfirmedTxs(context.Background(), &limit)
+	assert.NoError(err)
+	assert.NotNil(txRes)
+	assert.EqualValues(2, txRes.Count)
+	assert.EqualValues(2, txRes.Total)
+	assert.EqualValues(len(tx1)+len(tx2), txRes.TotalBytes)
+	assert.Len(txRes.Txs, 2)
+	assert.Contains(txRes.Txs, tx1)
+	assert.Contains(txRes.Txs, tx2)
+
+	// there's a bug in mempool implementation - count should be 1
+	// TODO(tzdybal): uncomment after resolving https://github.com/celestiaorg/optimint/issues/191
```

Comment:
I would also recommend revisting the tendermint mempool implementation. I think the tendermint team has been doing great improvements to the code base in the past few months.

Reply:
#191 is about using latest mempool from Tendermint - so as you pointed out - we want those improvements.